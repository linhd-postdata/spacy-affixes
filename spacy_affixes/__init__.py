# -*- coding: utf-8 -*-

"""Top-level package for SpaCy Affixes."""

__author__ = """LINHD POSTDATA Project"""
__email__ = 'info@linhd.uned.es'
__version__ = '0.1.0'
