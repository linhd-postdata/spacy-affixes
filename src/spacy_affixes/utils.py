import json
import os
import re
import sys
import unicodedata
from collections import defaultdict
from urllib.request import urlopen


AFFIXES_SUFFIX = "suffix"
AFFIXES_PREFIX = "prefix"
BASE_DIR = os.path.dirname(os.path.realpath(__file__))
DATA_DIR = os.path.join(BASE_DIR, "data")
FREELING_DIR = os.environ.get("FREELINGDIR") or os.environ.get("FREELINGSHARE")


def download(lang, version=None):
    version = version if version is not None else "4.1"
    affixes = download_affixes(lang, version)
    write_affixes(lang, version, affixes)
    lexicon = download_lexicon(lang, version)
    write_lexicon(lang, version, lexicon)


def write_affixes(lang, version, affixes):
    affixes_filename = f"affixes-{lang}-{version}.json"
    with open(os.path.join(DATA_DIR, affixes_filename), "w") as dump:
        json.dump(affixes, dump)


def write_lexicon(lang, version, lexicon):
    lexicon_filename = f"lexicon-{lang}-{version}.json"
    with open(os.path.join(DATA_DIR, lexicon_filename), "w") as dump:
        json.dump(lexicon, dump)


def load_affixes(lang="es", version="4.1"):
    affixes_filename = f"affixes-{lang}-{version}.json"
    affixes_path = os.path.join(DATA_DIR, affixes_filename)
    if not os.path.isfile(affixes_path):
        if FREELING_DIR:
            affixes_raw_path = os.path.join(FREELING_DIR, lang, "affixos.dat")
            with open(affixes_raw_path, "r") as affixes_raw:
                affixes = build_affixes(affixes_raw)
                write_affixes(lang, version, affixes)
                return affixes
        else:
            raise ValueError("""
            Data for affixes rules is missing. Check
            that Freeling is installed and its environment
            variable `FREELINGDIR` is set), or that you have downloaded the
            neccessary files using
            `python -m spacy_affixes download <lang> [version]`.
            Please, check the Freeling site to see license
            compatibilities.
            """)
    else:
        with open(affixes_path, "r") as dump:
            return json.load(dump)


def load_lexicon(lang="es", version="4.1"):
    lexicon_filename = f"lexicon-{lang}-{version}.json"
    lexicon_path = os.path.join(DATA_DIR, lexicon_filename)
    if not os.path.isfile(lexicon_path):
        if FREELING_DIR:
            lexicon_raw_path = os.path.join(FREELING_DIR, lang, "dicc.src")
            with open(lexicon_raw_path, "r") as lexicon_raw:
                lexicon = build_lexicon(lexicon_raw)
                write_lexicon(lang, version, lexicon)
                return lexicon
        else:
            raise ValueError("""
            Data for lexicon data is missing. Check
            that Freeling is installed and its environment
            variable `FREELINGDIR` is set), or that you have downloaded the
            neccessary files using
            `python -m spacy_affixes download <lang> [version]`.
            Please, check the Freeling site to see license
            compatibilities.
            """)
    else:
        with open(lexicon_path, "r") as dump:
            return json.load(dump)


def download_affixes(lang="es", version="4.1"):
    sys.stdout.write(f"Downloading affixes {lang}-{version}...\n")
    url = (f"https://raw.githubusercontent.com/TALP-UPC/FreeLing/"
           f"{version}/data/{lang}/afixos.dat")
    affixes_raw = urlopen(url).read().decode('utf-8')
    return build_affixes(affixes_raw)


def build_affixes(affixes_raw):
    re_flags = re.S | re.M | re.I
    prefixes_re = re.compile(r'<Prefixes.*>(.*)</Prefixes>', re_flags)
    suffixes_re = re.compile(r'<Suffixes.*>(.*)</Suffixes>', re_flags)
    prefixes = prefixes_re.search(affixes_raw).groups()[0].strip().split("\n")
    suffixes = suffixes_re.search(affixes_raw).groups()[0].strip().split("\n")
    affixes_rules = [(prefixes, AFFIXES_PREFIX, r"^{}"),
                     (suffixes, AFFIXES_SUFFIX, r"{}$")]
    affixes_dict = defaultdict(list)
    for (affixes, affix_kind, affix_pos_re) in affixes_rules:
        for affix in affixes:
            if len(affix.strip()) > 0 and not affix.startswith("#"):
                affix_split = re.split(r"\s+", affix)
                key, add, pos_re, _, strip_accent, *_, tokens = affix_split
                add = add if add != "*" else ""
                strip_accent = int(strip_accent) == 0
                if tokens == "-":
                    text = [key]
                else:
                    text = tokens.split(":")[0].split("+")[1:]
                rule = {
                    "pattern": affix_pos_re.format(key),
                    "kind": affix_kind,
                    "pos_re": fr"{pos_re}",
                    "strip_accent": strip_accent,
                    "affix_add": add.split("|"),
                    "affix_text": text,
                }
                affixes_dict[f"{affix_kind}_{key}"].append(rule)
    return affixes_dict


def strip_accents(string):
    return ''.join(char for char in unicodedata.normalize('NFD', string)
                   if (unicodedata.category(char) != 'Mn'
                       or unicodedata.name(char) == 'COMBINING TILDE'))


def eagle2tag(eagle):
    """Transform an EAGLE tag into UD features"""
    # TODO
    return ""


def eagle2pos(eagle):
    mapper = {
        "A": "ADJ",
        "R": "ADV",
        "D": "DET",
        "N": "NOUN",
        "V": "VERB",
        "P": "PRON",
        "C": "CONJ",
        "I": "INTJ",
        "S": "ADP",
        "F": "PUNCT",
        "Z": "NUM",
        "W": "NUM",  # Dates (W) are not standard EAGLE
    }
    return mapper.get(eagle[0], "X")


def token_transform(string, kind, add, strip_accent):
    if add == AFFIXES_PREFIX:
        prefix, suffix = add, ""
    else:
        prefix, suffix = "", add
    if add and strip_accent:
        transform = f"{prefix}{strip_accents(string)}{suffix}"
    elif add and not strip_accent:
        transform = f"{prefix}{string}{suffix}"
    elif not add and strip_accent:
        transform = strip_accents(string)
    else:
        transform = string
    return transform


def download_lexicon(lang="es", version="4.1"):
    sys.stdout.write(f"Downloading lexicon {lang}-{version}...\n")
    url = (f"https://raw.githubusercontent.com/TALP-UPC/FreeLing/"
           f"{version}/data/{lang}/dictionary/entries/MM.{{category}}")
    # Categories and Universal Dependency (UD) equivalent. None to auto-detect
    categories = (
        ("adj", "ADJ"),
        ("adv", "ADV"),
        ("int", "INTJ"),
        ("nom", "NOUN"),
        ("vaux", "AUX"),
        ("verb", "VERB"),
        ("tanc", None),
    )
    lexicon = defaultdict(list)
    for category, ud in categories:
        download_url = url.format(category=category)
        for line in urlopen(download_url).read().decode('utf-8').split("\n"):
            if len(line.strip()) == 0:
                continue
            word, lemma, eagle = line.split()
            lexicon[word].append({
                'lemma': lemma,
                'eagle': eagle,
                'ud': ud or eagle2pos(eagle),
                'tags': eagle2tag(eagle),
            })
    return lexicon


def build_lexicon(lexicon_raw):
    lexicon = defaultdict(list)
    for line in lexicon_raw.decode('utf-8').split("\n"):
        word, lemma, eagle = line.split()
        lexicon[word].append({
            'lemma': lemma,
            'eagle': eagle,
            'ud': eagle2pos(eagle),
            'tags': eagle2tag(eagle),
        })
    return lexicon


def get_morfo(string, lexicon, regex):
    # Checks for string in the lexicon
    # Returns EAGLE, UD, tags, lemma
    if string in lexicon:
        entry = lexicon[string]
        for definition in entry:
            if regex.match(definition["eagle"]):
                return (
                    definition["eagle"],
                    definition["ud"],
                    definition["tags"],
                    definition["lemma"]
                )
    return None
