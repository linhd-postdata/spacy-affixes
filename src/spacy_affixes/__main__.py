import sys
from .utils import download

USAGE = """Usage:
    python -m spacy_affixes download lang [version]

Parameters:
- lang. Two characters code for a language
- version. (Optional) Version to use (defaults to '4.1')

For example, the default behaviour is:
    python -m spacy_affixes download es 4.1
"""
# pragma: no cover
if __name__ == "__main__":
    argv_len = len(sys.argv)
    if 2 <= argv_len <= 4 and sys.argv[1] == "download":
        version = sys.argv[3] if argv_len == 4 else None
        download(lang=sys.argv[2], version=version)
    else:
        sys.stdout.write(USAGE)
