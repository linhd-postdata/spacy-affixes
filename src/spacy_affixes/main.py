# -*- coding: utf-8 -*-
"""Main module."""
import re

from spacy.matcher import Matcher
from spacy.tokens import Token

from .utils import AFFIXES_SUFFIX
from .utils import get_morfo
from .utils import load_affixes
from .utils import load_lexicon
from .utils import token_transform


class AffixesMatcher(object):

    def __init__(self, nlp, rules=None, lexicon=None, split_on=None):
        """
        :param nlp: SpaCy NLP object with the language already loaded
        :param rules: Dictionary of rules for affixes handling. Each dict
                      uses a key that contains the pattern to match and the
                      value is a list of dicts with the corresponding rule
                      parameters::
                      - pattern: Regular expression to match, (ex. `r"ito$"`)
                                 If a match is found, it gets removed from the
                                 token
                      - kind: `AFFIXES_SUFFIX` or `AFFIXES_PREFIX`
                      - pos_re: EAGLE regular expression to match, (ex. `r"V"`)
                      - strip_accent: Boolean indicating whether accents should
                                      be stripped in order to find the rest
                                      of the token in the lexicon,
                      - affix_add: List of strings to add to the rest of the
                                   token to find it in the lexicon. Each
                                   element in the list is tried separately, as
                                   in an OR condition. The character `*` means
                                   add nothing (ex. ["*", "io"])
                      - affix_text: List of Strings with the text to the rest
                                    of the token as individual tokens. For
                                    example, a rule for `dÃ­gamelo` might have
                                    ["me", "lo"] as its `affix_text`
                      It defaults to Freeling if installed (environment
                      variable `FREELINGDIR` should be set) or downloaded using
                      `python -m spacy_affixes download_freeling_data`. Please,
                      check the Freeling site to see license compatibilities.
        :param lexicon: Dictionary keyed by word with values for lemma,
                        EAGLE code, UD POS, and UD Tags. It defaults to
                        Freeling if installed (environment
                        variable `FREELINGDIR` should be set) or downloaded
                        using `python -m spacy_affixes download_freeling_data`.
                        Please, check the Freeling site to see license
                        compatibilities.
        :param split_on: Tuple of UD POS to split tokens on. Defaults to
                         verbs. A `*` means split whenever possible.
        """
        self.nlp = nlp
        self.rules = load_affixes() if rules is None else rules
        self.lexicon = load_lexicon() if lexicon is None else lexicon
        self.split_on = ("VERB", ) if split_on is None else split_on
        if None in (self.lexicon, self.rules):
            raise ValueError("""
            Data for affixes rules or lexicon data is missing. Check
            that Freeling is installed and its environment
            variable `FREELINGDIR` is set), or that you have downloaded the
            neccessary files using
            `python -m spacy_affixes download_freeling_data`.
            Please, check the Freeling site to see license
            compatibilities.
            """)
        if not Token.has_extension("has_affixes"):
            Token.set_extension("has_affixes", default=False)
            Token.set_extension("affixes_rule", default=None)
            Token.set_extension("affixes_text", default=None)
            Token.set_extension("affixes_kind", default=None)
            Token.set_extension("affixes_length", default=0)
        self.matcher = Matcher(nlp.vocab)
        for rule_key, rules in self.rules.items():
            for rule in rules:
                self.matcher.add(rule_key, None, [
                    {"TEXT": {"REGEX": fr"(?i){rule['pattern']}"}},
                    # It'd be nice if we could check regex AND minimum length
                    # {"LENGTH": {">": len(rule_key)}},
                ])

    def apply_rules(self, retokenizer, token, rule):
        for affix_add in rule["affix_add"]:
            token_sub = re.sub(rule["pattern"], '', token.text)
            token_left = token_transform(
                token_sub,
                rule["kind"],
                affix_add,
                rule["strip_accent"]
            )
            morfo = get_morfo(
                token_left.lower(),
                self.lexicon,
                re.compile(rule["pos_re"], re.I)
            )
            if (token_left and morfo and not token._.has_affixes):
                _, token_ud, token_tags, token_lemma = morfo
                affixes_length = (
                    len(rule["affix_text"]) or int(affix_add != "")
                )
                if rule["kind"] == AFFIXES_SUFFIX:
                    heads = [(token, 1)] + (affixes_length * [(token, 0)])
                    token.lemma_ = self.nlp.Defaults.lemma_lookup.get(
                        token_left.lower(),
                        token_lemma
                    )
                else:
                    heads = (affixes_length * [(token, 0)]) + [(token, 1)]
                if "*" in self.split_on or token_ud in self.split_on:
                    if rule["kind"] == AFFIXES_SUFFIX:
                        retokenizer.split(
                            token, [token_sub, *rule["affix_text"]], heads
                        )
                    else:
                        retokenizer.split(
                            token, [*rule["affix_text"], token_sub], heads
                        )
                token.pos_ = token_ud
                if token_tags:
                    token.tag_ = token_tags
                token._.affixes_text = token_left
                token._.affixes_kind = rule["kind"]
                token._.affixes_length = affixes_length
                token._.has_affixes = True

    def __call__(self, doc):
        matches = self.matcher(doc)
        tokens = []
        for match_id, start, end in matches:
            match_key = self.nlp.vocab.strings[match_id]
            for token in doc[start:end]:
                token._.affixes_rule = match_key
                tokens.append(token)
        with doc.retokenize() as retokenizer:
            for token in tokens:
                if token._.affixes_rule:
                    for rule in self.rules[token._.affixes_rule]:
                        self.apply_rules(retokenizer, token, rule)
        return doc
