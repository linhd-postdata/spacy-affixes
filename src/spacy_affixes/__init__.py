# -*- coding: utf-8 -*-
# flake8: noqa
"""Top-level package for SpaCy Affixes."""

__author__ = """LINHD POSTDATA Project"""
__email__ = 'info@linhd.uned.es'
__version__ = '0.1.1'

from .main import AffixesMatcher  # pragma: no cover
