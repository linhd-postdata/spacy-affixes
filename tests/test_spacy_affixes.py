#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Tests for `spacy_affixes` package."""
import pytest
import spacy

from spacy_affixes import AffixesMatcher
from spacy_affixes.utils import download


@pytest.fixture
def nlp():
    lang = "es"
    download(lang)
    nlp_ = spacy.load(lang)
    if nlp_.has_pipe("affixes"):
        nlp_.remove_pipe("affixes")
    return nlp_

def test_split_on_all(snapshot, nlp):
    affixes_matcher = AffixesMatcher(nlp, split_on='*')
    nlp.add_pipe(affixes_matcher, name="affixes", before="tagger")
    docs = (
        "Cuéntamelo bien y dilo claro, no me des un caramelo.",
        "Yo mismamente podría hacérselo despacito.",
        "Soy hispanoamericano y antirrevolucionario.",
        "Dime el número de teléfono.",
        "Hay que hacérselo todo.",
    )
    for doc in docs:
        snapshot.assert_match([
        [
            token.text,
            token.lemma_,
            token.pos_,
            token.tag_,
            token._.has_affixes,
            token._.affixes_rule,
            token._.affixes_kind,
            token._.affixes_text,
            token._.affixes_length,
        ] for token in nlp(doc)])

def test_split_on_verbs(snapshot, nlp):
    affixes_matcher = AffixesMatcher(nlp, split_on=["VERB"])
    nlp.add_pipe(affixes_matcher, name="affixes", before="tagger")
    docs = (
        "Cuéntamelo bien y dilo claro, no me des un caramelo.",
        "Yo mismamente podría hacérselo despacito.",
        "Soy hispanoamericano y antirrevolucionario.",
        "Dime el número de teléfono.",
        "Hay que hacérselo todo.",
    )
    for doc in docs:
        snapshot.assert_match([
        [
            token.text,
            token.lemma_,
            token.pos_,
            token.tag_,
            token._.has_affixes,
            token._.affixes_rule,
            token._.affixes_kind,
            token._.affixes_text,
            token._.affixes_length,
        ] for token in nlp(doc)])
