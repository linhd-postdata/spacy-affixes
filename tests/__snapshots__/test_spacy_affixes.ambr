# serializer version: 1
# name: test_accent_exceptions
  list([
    list([
      'Ese',
      'ese',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hombre',
      'hombre',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'está',
      'estar',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'demente',
      'demente',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_accent_exceptions.1
  list([
    list([
      'automáticamente',
      'automático',
      True,
      'suffix_mente',
      'suffix',
      'automática',
      1,
    ]),
  ])
# ---
# name: test_accent_exceptions.2
  list([
    list([
      'mágicamente',
      'mágicamente',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_accent_exceptions.3
  list([
    list([
      'antirrevolucionariamente',
      'antirrevolucionario',
      True,
      'suffix_mente',
      'suffix',
      'antirrevolucionaria',
      1,
    ]),
  ])
# ---
# name: test_split_on_all
  list([
    list([
      'Cuénta',
      'contar',
      'NOUN',
      'NOUN',
      True,
      'suffix_melo',
      'suffix',
      'Cuenta',
      2,
    ]),
    list([
      'me',
      'me',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'bien',
      'bien',
      'ADV',
      'ADV',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'y',
      'y',
      'CCONJ',
      'CCONJ',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'di',
      'decir',
      'NOUN',
      'NOUN',
      True,
      'suffix_lo',
      'suffix',
      'di',
      1,
    ]),
    list([
      'lo',
      'lo',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'claro',
      'claro',
      'ADJ',
      'ADJ',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      ',',
      ',',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'no',
      'no',
      'ADV',
      'ADV',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'me',
      'yo',
      'PRON',
      'PRON',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'des',
      'des',
      'ADP',
      'ADP',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'un',
      'uno',
      'DET',
      'DET',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'caramelo',
      'caramelo',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_all.1
  list([
    list([
      'Yo',
      'yo',
      'PRON',
      'PRON',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'misma',
      'mismo',
      'NOUN',
      'NOUN',
      True,
      'suffix_mente',
      'suffix',
      'misma',
      1,
    ]),
    list([
      'mente',
      'mente',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'podría',
      'poder',
      'AUX',
      'AUX',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hacér',
      'hacer',
      'NOUN',
      'NOUN',
      True,
      'suffix_selo',
      'suffix',
      'hacer',
      2,
    ]),
    list([
      'se',
      'se',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'despac',
      'despacio',
      'NOUN',
      'NOUN',
      True,
      'suffix_ito',
      'suffix',
      'despacio',
      1,
    ]),
    list([
      'ito',
      'ito',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_all.2
  list([
    list([
      'Soy',
      'ser',
      'AUX',
      'AUX',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hispano',
      'hispano',
      'NOUN',
      'NOUN',
      True,
      'prefix_hispano',
      'prefix',
      'americano',
      1,
    ]),
    list([
      'americano',
      'americano',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'y',
      'y',
      'CCONJ',
      'CCONJ',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'antirrevolucionario',
      'antirrevolucionario',
      'ADJ',
      'ADJ',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_all.3
  list([
    list([
      'Di',
      'decir',
      'NOUN',
      'NOUN',
      True,
      'suffix_me',
      'suffix',
      'Di',
      1,
    ]),
    list([
      'me',
      'me',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'el',
      'el',
      'DET',
      'DET',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'número',
      'número',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'de',
      'de',
      'ADP',
      'ADP',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'teléfono',
      'teléfono',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_all.4
  list([
    list([
      'Hay',
      'haber',
      'AUX',
      'AUX',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'que',
      'que',
      'SCONJ',
      'SCONJ',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hacér',
      'hacer',
      'NOUN',
      'NOUN',
      True,
      'suffix_selo',
      'suffix',
      'hacer',
      2,
    ]),
    list([
      'se',
      'se',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      'NOUN',
      'NOUN',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'todo',
      'todo',
      'PRON',
      'PRON',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      'PUNCT',
      'PUNCT',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_verbs
  list([
    list([
      'Cuénta',
      'contar',
      True,
      'suffix_melo',
      'suffix',
      'Cuenta',
      2,
    ]),
    list([
      'me',
      'me',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'bien',
      'bien',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'y',
      'y',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'di',
      'decir',
      True,
      'suffix_lo',
      'suffix',
      'di',
      1,
    ]),
    list([
      'lo',
      'lo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'claro',
      'claro',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      ',',
      ',',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'no',
      'no',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'me',
      'yo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'des',
      'des',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'un',
      'uno',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'caramelo',
      'caramelo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_verbs.1
  list([
    list([
      'Yo',
      'yo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'mismamente',
      'mismo',
      True,
      'suffix_mente',
      'suffix',
      'misma',
      1,
    ]),
    list([
      'podría',
      'poder',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hacér',
      'hacer',
      True,
      'suffix_selo',
      'suffix',
      'hacer',
      2,
    ]),
    list([
      'se',
      'se',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'despacito',
      'despacio',
      True,
      'suffix_ito',
      'suffix',
      'despacio',
      1,
    ]),
    list([
      '.',
      '.',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_verbs.2
  list([
    list([
      'Soy',
      'ser',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hispanoamericano',
      'hispanoamericano',
      True,
      'prefix_hispano',
      'prefix',
      'americano',
      1,
    ]),
    list([
      'y',
      'y',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'antirrevolucionario',
      'antirrevolucionario',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_verbs.3
  list([
    list([
      'Di',
      'decir',
      True,
      'suffix_me',
      'suffix',
      'Di',
      1,
    ]),
    list([
      'me',
      'me',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'el',
      'el',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'número',
      'número',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'de',
      'de',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'teléfono',
      'teléfono',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
# name: test_split_on_verbs.4
  list([
    list([
      'Hay',
      'haber',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'que',
      'que',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'hacér',
      'hacer',
      True,
      'suffix_selo',
      'suffix',
      'hacer',
      2,
    ]),
    list([
      'se',
      'se',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'lo',
      'lo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      'todo',
      'todo',
      False,
      None,
      None,
      None,
      0,
    ]),
    list([
      '.',
      '.',
      False,
      None,
      None,
      None,
      0,
    ]),
  ])
# ---
